syntax = "proto3";

import "google/protobuf/wrappers.proto";
import "types.proto";

package featureflag.services.types;

option csharp_namespace = "FeatureFlag.Services.Types";

// * 1.0.0 - Initial contract implementation
// * 1.1.0 - Support authentication for SetFeatureFlag method
// * 1.2.0 - Add DeleteFeatureFlag method
// * 1.3.0 - Add SetFeatureFlagForStations method
// * 1.3.1 - Move version property to types.proto file
// * 1.4.0 - Add IsFeatureFlagEnabledForStations method
option (version) = "1.4.0";

/*
The Feature Flag Service provides methods to get station feature flag values

Authentication
^^^^^^^^^^^^^^
Please provide a valid ``api_key`` parameter in gRPC service metadata for ``SetFeatureFlag`` and ``DeleteFeatureFlag`` method, otherwise the service will return ``Unauthenticated`` response.
*/ 
service FeatureFlagService {
  // Gets Feature Flag information
  rpc GetFeatureFlag(GetFeatureFlagRequest) returns (GetFeatureFlagResponse);

  // Updates feature flag
  rpc SetFeatureFlag(SetFeatureFlagRequest) returns (SetFeatureFlagResponse);

  // Delete a feature flag for a given application name
  rpc DeleteFeatureFlag(DeleteFeatureFlagRequest) returns (DeleteFeatureFlagResponse);

  // Set feature flag for multiple stations
  rpc SetFeatureFlagForStations(SetFeatureFlagForStationsRequest) returns (SetFeatureFlagForStationsResponse);

  // Get feature flag value for multiple stations (maximum number of stations is 500)
  rpc IsFeatureFlagEnabledForStations(IsFeatureFlagEnabledForStationsRequest) returns (IsFeatureFlagEnabledForStationsResponse);
}

message SetFeatureFlagRequest {
	string station_external_id = 1;
	string application_name = 2;
	string feature_name = 3;
	bool feature_enabled = 4;
}

message SetFeatureFlagResponse {
}

message GetFeatureFlagRequest {
	string station_external_id = 1;
	string application_name = 2;
	string feature_name = 3;
}

message GetFeatureFlagResponse {
	string feature_flag_external_id = 1;
	bool feature_enabled = 2;
}

// Delete feature flag request message
message DeleteFeatureFlagRequest {
	string feature_name = 1;
	string application_name = 2;
  // this parameter is optional to support deleting one feature flag for all stations
	google.protobuf.StringValue station_external_id = 3;
}

// Delete feature flag response message
message DeleteFeatureFlagResponse {
}

message SetFeatureFlagForStationsRequest {
	repeated string station_external_ids = 1;
	string application_name = 2;
	string feature_name = 3;
	bool feature_enabled = 4;
}

message SetFeatureFlagForStationsResponse {
}

message IsFeatureFlagEnabledForStationsRequest {
	repeated string station_external_ids = 1;
	string application_name = 2;
	string feature_name = 3;
}

message IsFeatureFlagEnabledForStationsResponse {
	repeated string station_external_ids = 1;
}
